### PR TITLE
Obtain initial access_token if necessary in BatchHttpRequest.execute()

### LINKED ISSUE #211: BatchHttpRequest does not check for existence of an accesstoken before calling apply()
@nathanielmanistaatgoogle @jonparrott 

To repro:

```
from googleapiclient.discovery import build
from oauth2client.client import GoogleCredentials

def foo(b, a, r):
   pass

def main():
    api = build('servicemanagement', 'v1', credentials=GoogleCredentials.from_stream('static/service_account.json'))
    batch = api.new_batch_http_request(callback=foo)
    batch.add(api.services().list())
    batch.execute()
```

You will get something like 

  File "/usr/local/google/home/elibixby/gobrepos/devrel/apps/cloud/free-trial-tour/project_utils/apis.py", line 45, in enable_async
    batch.execute()
  File "/usr/local/google/home/elibixby/envs/iliad/local/lib/python2.7/site-packages/oauth2client/util.py", line 137, in positional_wrapper
    return wrapped(*args, **kwargs)
  File "/usr/local/google/home/elibixby/envs/iliad/local/lib/python2.7/site-packages/googleapiclient/http.py", line 1325, in execute
    self._execute(http, self._order, self._requests)
  File "/usr/local/google/home/elibixby/envs/iliad/local/lib/python2.7/site-packages/googleapiclient/http.py", line 1249, in _execute
    body = self._serialize_request(request)
  File "/usr/local/google/home/elibixby/envs/iliad/local/lib/python2.7/site-packages/googleapiclient/http.py", line 1120, in _serialize_request
    request.http.request.credentials.apply(headers)
  File "/usr/local/google/home/elibixby/envs/iliad/local/lib/python2.7/site-packages/oauth2client/client.py", line 676, in apply
    headers['Authorization'] = 'Bearer ' + self.access_token
TypeError: cannot concatenate 'str' and 'NoneType' objects

FYI my pip freeze is:
```
Flask==0.10.1
Flask-SeaSurf==0.2.2
Jinja2==2.8
MarkupSafe==0.23
Werkzeug==0.11.5
argparse==1.2.1
google-api-python-client==1.5.0
httplib2==0.9.2
itsdangerous==0.24
oauth2client==2.0.1
pyasn1==0.1.9
pyasn1-modules==0.0.8
requirements-dev.txt==0.0.0
rsa==3.4.2
simplejson==3.8.2
six==1.10.0
talisman==0.1.0
uritemplate==0.6
wsgiref==0.1.2
```

### PR DESCRIPTION
**Pull request by Garciat**
Fixes #211

### CODE REVIEW

**Comment by jonparrott** [comment]
Thanks for your contribution. Can you add tests to cover this case?

**Comment by Garciat** [comment]
I've implemented a test. Writing it actually made me realize that the fix is not exactly ideal, since different type of `Credentials`-derived object may exist which are not access token-based. A more involved fix would require actually providing a method in the base `Credentials` class akin to `is_ready()` or whatever.

File: googleapiclient/http.py [hunk 1]
```diff
@@ -1394,6 +1394,15 @@ def execute(self, http=None):
     if http is None:
       raise ValueError("Missing a valid http object.")
 
+    # Special case for OAuth2Credentials-style objects which have not yet been
+    # refreshed with an initial access_token.
+    if http is not None and hasattr(http.request, 'credentials'):
```
Comment (nathanielmanistaatgoogle) [inline googleapiclient/http.py]: [Never `hasattr`](http://doc.bazaar.canonical.com/developers/code-style.html#hasattr-and-getattr).

File: googleapiclient/http.py [hunk 2]
```diff
@@ -1394,6 +1394,15 @@ def execute(self, http=None):
...
+      creds = http.request.credentials
+      if hasattr(creds, 'access_token') and not creds.access_token:
```
Comment (nathanielmanistaatgoogle) [inline googleapiclient/http.py]: [Never `hasattr`](http://doc.bazaar.canonical.com/developers/code-style.html#hasattr-and-getattr).

File: googleapiclient/http.py [hunk 3]
```diff
@@ -1394,6 +1394,15 @@ def execute(self, http=None):
...
+        LOGGER.info('Attempting refresh to obtain '
```
Comment (nathanielmanistaatgoogle) [inline googleapiclient/http.py]: Please avoid implicit string concatenation whenever you can - does this string really not fit in eighty columns? If no, line-wrap after the opening parenthesis?

File: tests/test_http.py [hunk 4]
```diff
@@ -1072,6 +1072,34 @@ def test_execute_request_body(self):
...
+       BATCH_SINGLE_RESPONSE),
+    ])
+
+    cred.authorize(http)
+
+    batch.add(self.request1, callback=callbacks.f)
+    batch.execute(http=http)
+
+    self.assertEqual({'foo': 42}, callbacks.responses['1'])
+    self.assertEqual(None, callbacks.exceptions['1'])
```
Comment (nathanielmanistaatgoogle) [inline tests/test_http.py]: `assertIsNone` rather than `assertEqual` with one `None` argument.

**Comment by nathanielmanistaatgoogle** [comment]
Please squash commits - there's no need for "Revert local testing changes." to be part of canonical project history.

**Comment by Garciat** [comment]
Had a small confusion committing from a different computer. I pushed a commit with wrong email information, and the bot removed the cla tag. Said commit has been discarded now. I'll try commenting "I signed it!" below.

**Comment by jonparrott** [comment]
> Please squash commits - there's no need for "Revert local testing changes." to be part of canonical project history.

@nathanielmanistaatgoogle we can do that when we merge now. :)

**Comment by nathanielmanistaatgoogle** [comment]
One last small request: please edit your commit log message to conform to [these guidelines](http://chris.beams.io/posts/git-commit/#seven-rules).

**Comment by Garciat** [comment]
I've updated the commit message to adhere to the guidelines.

**Comment by nathanielmanistaatgoogle** [comment]
Thank you for your contribution!

**Comment by Garciat** [comment]
Thank you ! Learned a few things (:

**Pull request merged**